### DISCUSSION TITLE
프록시 패턴

### DISCUSSION THREAD

**Discussion by jangwooooo** [subak-ba/cs-study#16] [디자인 패턴]
## 프록시 패턴
프록시 패턴이란 어떤 대상 객체에 접근하기 전 그 접근에 대한 흐름을 가로채서 추가적인 로직을 처리한 뒤 접근하게 하는 디자인 패턴이다.    

<br/>

## 패턴 종류
- 가상 프록시
  - 객체 생성에 많은 자원이 소모되지만 사용 빈도는 적을 때 사용된다. (지연 초기화) 
- 보호 프록시
  - 특정 권한만 객체를 사용할 수 있도록 할 때 사용된다. 
- 로깅 프록시
  - 대상 객체에 대한 로깅을 위해 사용된다 
- 캐싱 프록시
  - 요청 결과를 캐시하고 캐시의 생명주기를 관리할 때 사용된다.
- 원격 프록시
  - 프록시 클래스는 로컬에 있고, 대상 객체는 원격에 존재할 때 사용된다.

<br/>

## 패턴 사용 시기
- 접근을 제어하거가 기능을 추가하고 싶은데, 기존의 특정 객체를 수정할 수 없는 상황일때
- 초기화 지연, 접근 제어, 로깅, 캐싱 등 기존 객체 동작에 수정 없이 가미하고 싶을 때

<br/>

## 장점
- 개방폐쇄원칙(OCP) 준수 
  - 기존 대상 객체의 코드를 변경하지 않고 새로운 기능을 추가할 수 있다.
- 단일책임원칙(SRP) 준수
  - 대상 객체는 자신의 기능에만 집중 하고 그 이외 부가 기능을 제공하는 역할을 프록시 객체에 위임하여 다중 책임을 회피 할 수 있다.
- 클라이언트는 객체를 신경쓰지 않고 서비스 객체를 제어할 수 있다.
- 사용자 입장에서는 프록시 객체나 실제 객체나 사용법은 유사하여 사용성에 문제 되지 않는다.

## 단점
- 여러 클래스에 기능을 가미 시키고 싶다면 많은 프록시 클래스를 도입해야 하므로 코드의 복잡도가 증가한다.
- 프록시 클래스 자체에 들어가는 자원이 많다면 서비스로부터의 응답이 늦어질 수 있다.

#4

**Comment by onee-only** [comment]
이해를 돕기 위한 사진도 추가 해 주실 수 있나요?
